### PR TITLE
[chip-tool] Use an onDone callback when reading events

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

`./out/debug/standalone/chip-tool testcluster read-event test-event 0x12345 1` waits forever if no events exists.
This is because the `ReadEvent` API does not inform the caller of the completion of the request.

#### Change overview
 * Add a `onDoneCb` argument
 * Improve logging by passing the event header up to the success/error callbacks
 
#### Testing
`./out/debug/standalone/chip-tool testcluster read-event test-event 0x12345 1`  now returns.

### CODE REVIEW

File: examples/chip-tool/templates/commands.zapt [hunk 1]
```diff
@@ -36,6 +36,30 @@ CHIP_ERROR LogValue(const char * label, size_t indent, {{zapTypeToDecodableClust
 {{/zcl_structs}}
 {{/zcl_clusters}}
 
+void LogEventHeader(const chip::app::EventHeader & eventHeader)
```
Comment (andy31415) [inline examples/chip-tool/templates/commands.zapt:39]: This function does not seem to depend on  parameters  given to the template. Please move to a propper C++/header as there is no reason for this to be "codegen".

File: examples/chip-tool/templates/commands.zapt [hunk 2]
```diff
@@ -217,14 +241,20 @@ static void OnDefaultSuccess(void * context, const chip::app::DataModel::NullObj
...
-static void OnGeneralAttributeEventResponse(void * context, const char * label, T value)
+static void OnGeneralAttributeResponse(void * context, const char * label, T value)
 {
     CHIP_ERROR err = LogValue(label, 0, value);
 
     auto * command = static_cast<ModelCommand *>(context);
     command->SetCommandExitStatus(err);
 }
 
+static void OnDefaultEventFailure(void * context, const chip::app::EventHeader & eventHeader, EmberAfStatus status)
```
Comment (andy31415) [inline examples/chip-tool/templates/commands.zapt:252]: No need to codegen this either.

**Review by andy31415 — CHANGES_REQUESTED** [review]
Changes requested for not adding non-generated code to codegen.

Complex codegen is harder to  review and maintain: no syntax highlighting, C++ tools will not go to the actual definition (will go to codegen if that is not hidden away), cannot be unit tested easily, is duplicated in many places.

File: src/controller/ReadInteraction.h [hunk 3]
```diff
@@ -218,11 +224,13 @@ CHIP_ERROR ReportEvent(Messaging::ExchangeManager * apExchangeMgr, EndpointId en
 template <typename DecodableEventType>
 CHIP_ERROR ReadEvent(Messaging::ExchangeManager * exchangeMgr, const SessionHandle & sessionHandle, EndpointId endpointId,
                      typename TypedReadEventCallback<DecodableEventType>::OnSuccessCallbackType onSuccessCb,
-                     typename TypedReadEventCallback<DecodableEventType>::OnErrorCallbackType onErrorCb)
+                     typename TypedReadEventCallback<DecodableEventType>::OnErrorCallbackType onErrorCb,
+                     typename TypedReadEventCallback<DecodableEventType>::OnSubscriptionEstablishedCallbackType onDoneCb = nullptr)
```
Comment (mrjerryjohns) [inline src/controller/ReadInteraction.h:228]: If the expectation of this function is that you're trying to read out a specific type of event, and if you don't actually get any of those events, shouldn't that be interpreted as an error (in the context of this function)?

If so, I don't think adding a onDone cb is quite right - rather, it should call `onErrorCb`.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

**Comment by vivien-apple** [comment]
Closing this one. I have an other approach.

**Pull request closed without merging**